cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -27,19 +32,139 @@
  */
 public class BasicWorkflowToken implements WorkflowToken {
   private Map<String, Map<String, Long>> mapReduceCounters;
+  private Map<String, List<NodeValueEntry>> tokenValueMap = Maps.newHashMap();
+  private String nodeName;
+
+  void setCurrentNode(String nodeName) {
+    this.nodeName = nodeName;
+  }
+
+  /**
+   * Merge the other WorkflowToken passed to the method as a parameter
+   * with the WorkflowToken on which the method is invoked.
+   * @param other the other WorkflowToken to be merged
+   */
+  void mergeToken(BasicWorkflowToken other) {
+    Map<String, List<NodeValueEntry>> otherTokenValueMap = other.tokenValueMap;
+
+    for (Map.Entry<String, List<NodeValueEntry>> entry : otherTokenValueMap.entrySet()) {
+      if (!tokenValueMap.containsKey(entry.getKey())) {
+        // Key is newly added to the other WorkflowToken
+        tokenValueMap.put(entry.getKey(), entry.getValue());
```

Comment:
The other list may not be immutable. It's not safe to just use the List returned by getValue. Also, after the merge, you may append new entries to the List right? Meaning it will modify the List inside another object?